File: win/ClipboardSyncWin/CryptoBox.cs
using System;
using System.Collections.Concurrent;
using System.Security.Cryptography;
using System.Text;

namespace ClipboardSyncWin;

internal static class CryptoBox
{
    private const string EnvelopeType = "encrypted";
    private const int EnvelopeVersion = 1;
    private const int RealtimeEnvelopeVersion = 2;
    private const int SaltBytes = 16;
    private const int NonceBytes = 12;
    private const int TagBytes = 16;
    private const int KeyBytes = 32;
    private const int Pbkdf2Rounds = 100_000;
    private static readonly byte[] RealtimeSalt = Encoding.UTF8.GetBytes("ClipboardSync realtime input v1");
    private static readonly byte[] SignedSalt = Encoding.UTF8.GetBytes("ClipboardSync signed transport v1");
    private static readonly ConcurrentDictionary<string, byte[]> KeyCache = new();

    public static EncryptedEnvelope Encrypt(byte[] plaintext, string password)
    {
        var salt = RandomNumberGenerator.GetBytes(SaltBytes);
        var nonce = RandomNumberGenerator.GetBytes(NonceBytes);
        var key = DeriveKey(password, salt);
        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[TagBytes];

        using var aes = new AesGcm(key, TagBytes);
        aes.Encrypt(nonce, plaintext, ciphertext, tag);

        return new EncryptedEnvelope
        {
            Type = EnvelopeType,
            Version = EnvelopeVersion,
            Salt = Convert.ToBase64String(salt),
            Nonce = Convert.ToBase64String(nonce),
            Ciphertext = Convert.ToBase64String(ciphertext),
            Tag = Convert.ToBase64String(tag)
        };
    }

    public static EncryptedEnvelope EncryptRealtime(byte[] plaintext, string password)
    {
        var nonce = RandomNumberGenerator.GetBytes(NonceBytes);
        var key = DeriveCachedKey(password, RealtimeSalt);
        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[TagBytes];

        using var aes = new AesGcm(key, TagBytes);
        aes.Encrypt(nonce, plaintext, ciphertext, tag);

        return new EncryptedEnvelope
        {
            Type = EnvelopeType,
            Version = RealtimeEnvelopeVersion,
            Salt = Convert.ToBase64String(RealtimeSalt),
            Nonce = Convert.ToBase64String(nonce),
            Ciphertext = Convert.ToBase64String(ciphertext),
            Tag = Convert.ToBase64String(tag)
        };
    }

    public static byte[] Decrypt(EncryptedEnvelope envelope, string password)
    {
        if (envelope.Type != EnvelopeType ||
            (envelope.Version != EnvelopeVersion && envelope.Version != RealtimeEnvelopeVersion))
        {
            throw new CryptographicException("Unsupported encrypted message.");
        }

        var salt = Convert.FromBase64String(envelope.Salt);
        var nonce = Convert.FromBase64String(envelope.Nonce);
        var ciphertext = Convert.FromBase64String(envelope.Ciphertext);
        var tag = Convert.FromBase64String(envelope.Tag);
        var plaintext = new byte[ciphertext.Length];
        var key = envelope.Version == RealtimeEnvelopeVersion
            ? DeriveCachedKey(password, salt)
            : DeriveKey(password, salt);

        using var aes = new AesGcm(key, TagBytes);
        aes.Decrypt(nonce, ciphertext, tag, plaintext);
        return plaintext;
    }

    /// Authenticated-plaintext transport: the payload travels unencrypted but
    /// carries an HMAC-SHA256 keyed by a password-derived key, so the password
    /// still authenticates every message when transport encryption is off.
    public static SignedEnvelope Sign(byte[] plaintext, string password)
    {
        var key = DeriveCachedKey(password, SignedSalt);
        using var hmac = new HMACSHA256(key);
        return new SignedEnvelope
        {
            Payload = Convert.ToBase64String(plaintext),
            Mac = Convert.ToBase64String(hmac.ComputeHash(plaintext))
        };
    }

    public static byte[] Verify(SignedEnvelope envelope, string password)
    {
        if (envelope.Type != "signed" || envelope.Version != 1)
        {
            throw new CryptographicException("Unsupported signed message.");
        }
        var plaintext = Convert.FromBase64String(envelope.Payload);
        var expectedMac = Convert.FromBase64String(envelope.Mac);
        var key = DeriveCachedKey(password, SignedSalt);
        using var hmac = new HMACSHA256(key);
        if (!CryptographicOperations.FixedTimeEquals(hmac.ComputeHash(plaintext), expectedMac))
        {
            throw new CryptographicException("Signed message authentication failed.");
        }
        return plaintext;
    }

    // Raw framing primitives.
    //
    // These skip the JSON envelope entirely: no base64 of the plaintext, no base64 of the
    // ciphertext, no JSON encode/decode. TunnelFrame uses them to put port-forward payloads on the
    // wire as raw bytes. They share the cached realtime/signed keys with the envelope paths, so
    // there is no extra PBKDF2 work.

    public const int RawNonceBytes = NonceBytes;
    public const int RawTagBytes = TagBytes;
    public const int RawMacBytes = 32;

    /// AES-GCM over raw bytes, with associatedData (the frame header) authenticated but not
    /// encrypted, so a relay cannot rewrite the connection id or routing fields without the tag
    /// failing.
    public static void SealRaw(
        ReadOnlySpan<byte> plaintext,
        ReadOnlySpan<byte> associatedData,
        string password,
        Span<byte> nonce,
        Span<byte> tag,
        Span<byte> ciphertext)
    {
        RandomNumberGenerator.Fill(nonce);
        var key = DeriveCachedKey(password, RealtimeSalt);
        using var aes = new AesGcm(key, TagBytes);
        aes.Encrypt(nonce, plaintext, ciphertext, tag, associatedData);
    }

    /// Throws CryptographicException when the tag does not verify.
    public static byte[] OpenRaw(
        ReadOnlySpan<byte> nonce,
        ReadOnlySpan<byte> tag,
        ReadOnlySpan<byte> ciphertext,
        ReadOnlySpan<byte> associatedData,
        string password)
    {
        if (nonce.Length != NonceBytes || tag.Length != TagBytes)
        {
            throw new CryptographicException("Malformed tunnel frame.");
        }
        var key = DeriveCachedKey(password, RealtimeSalt);
        var plaintext = new byte[ciphertext.Length];
        using var aes = new AesGcm(key, TagBytes);
        aes.Decrypt(nonce, ciphertext, tag, plaintext, associatedData);
        return plaintext;
    }

    /// HMAC-SHA256 for the authenticated-plaintext mode, keyed the same way as Sign.
    public static byte[] MacRaw(ReadOnlySpan<byte> data, string password)
    {
        var key = DeriveCachedKey(password, SignedSalt);
        var mac = new byte[RawMacBytes];
        HMACSHA256.HashData(key, data, mac);
        return mac;
    }

    public static bool IsValidMacRaw(ReadOnlySpan<byte> mac, ReadOnlySpan<byte> data, string password)
    {
        return CryptographicOperations.FixedTimeEquals(MacRaw(data, password), mac);
    }

    private static byte[] DeriveCachedKey(string password, byte[] salt)
    {
        var cacheKey = $"{password}\0{Convert.ToBase64String(salt)}";
        return KeyCache.GetOrAdd(cacheKey, _ => DeriveKey(password, salt));
    }

    private static byte[] DeriveKey(string password, byte[] salt)
    {
        return Rfc2898DeriveBytes.Pbkdf2(
            Encoding.UTF8.GetBytes(password),
            salt,
            Pbkdf2Rounds,
            HashAlgorithmName.SHA256,
            KeyBytes);
    }
}

File: mac/Sources/ClipboardSyncMac/CryptoBox.swift
import CommonCrypto
import CryptoKit
import Foundation

enum CryptoBox {
    private static let type = "encrypted"
    private static let version = 1
    private static let realtimeVersion = 2
    private static let saltBytes = 16
    private static let nonceBytes = 12
    private static let tagBytes = 16
    private static let keyBytes = 32
    private static let pbkdf2Rounds = 100_000
    private static let realtimeSalt = Data("ClipboardSync realtime input v1".utf8)
    private static let signedSalt = Data("ClipboardSync signed transport v1".utf8)
    private static let keyCacheQueue = DispatchQueue(label: "ClipboardSyncMac.crypto.keyCache")
    private static var cachedKeys: [String: SymmetricKey] = [:]

    static func encrypt(_ plaintext: Data, password: String) throws -> EncryptedEnvelope {
        let salt = randomData(count: saltBytes)
        let nonceData = randomData(count: nonceBytes)
        let key = try deriveKey(password: password, salt: salt)
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sealedBox = try AES.GCM.seal(plaintext, using: key, nonce: nonce)

        return EncryptedEnvelope(
            type: type,
            version: version,
            salt: salt.base64EncodedString(),
            nonce: nonceData.base64EncodedString(),
            ciphertext: sealedBox.ciphertext.base64EncodedString(),
            tag: sealedBox.tag.base64EncodedString()
        )
    }

    static func encryptRealtime(_ plaintext: Data, password: String) throws -> EncryptedEnvelope {
        let nonceData = randomData(count: nonceBytes)
        let key = try cachedKey(password: password, salt: realtimeSalt)
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sealedBox = try AES.GCM.seal(plaintext, using: key, nonce: nonce)

        return EncryptedEnvelope(
            type: type,
            version: realtimeVersion,
            salt: realtimeSalt.base64EncodedString(),
            nonce: nonceData.base64EncodedString(),
            ciphertext: sealedBox.ciphertext.base64EncodedString(),
            tag: sealedBox.tag.base64EncodedString()
        )
    }

    /// Authenticated-plaintext transport: the payload travels unencrypted but
    /// carries an HMAC-SHA256 keyed by a password-derived key, so the password
    /// still authenticates every message when transport encryption is off.
    static func sign(_ plaintext: Data, password: String) throws -> SignedEnvelope {
        let key = try cachedKey(password: password, salt: signedSalt)
        let mac = HMAC<SHA256>.authenticationCode(for: plaintext, using: key)
        return SignedEnvelope(
            type: "signed",
            version: 1,
            payload: plaintext.base64EncodedString(),
            mac: Data(mac).base64EncodedString(),
            from: nil,
            to: nil
        )
    }

    static func verify(_ envelope: SignedEnvelope, password: String) throws -> Data {
        guard envelope.type == "signed", envelope.version == 1 else {
            throw CryptoError.unsupportedEnvelope
        }
        guard
            let plaintext = Data(base64Encoded: envelope.payload),
            let mac = Data(base64Encoded: envelope.mac)
        else {
            throw CryptoError.invalidEnvelope
        }
        let key = try cachedKey(password: password, salt: signedSalt)
        guard HMAC<SHA256>.isValidAuthenticationCode(mac, authenticating: plaintext, using: key) else {
            throw CryptoError.invalidEnvelope
        }
        return plaintext
    }

    static func decrypt(_ envelope: EncryptedEnvelope, password: String) throws -> Data {
        guard envelope.type == type, envelope.version == version || envelope.version == realtimeVersion else {
            throw CryptoError.unsupportedEnvelope
        }
        guard
            let salt = Data(base64Encoded: envelope.salt),
            let nonceData = Data(base64Encoded: envelope.nonce),
            let ciphertext = Data(base64Encoded: envelope.ciphertext),
            let tag = Data(base64Encoded: envelope.tag),
            tag.count == tagBytes
        else {
            throw CryptoError.invalidEnvelope
        }

        let key = envelope.version == realtimeVersion
            ? try cachedKey(password: password, salt: salt)
            : try deriveKey(password: password, salt: salt)
        let nonce = try AES.GCM.Nonce(data: nonceData)
        let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        return try AES.GCM.open(sealedBox, using: key)
    }

    private static func cachedKey(password: String, salt: Data) throws -> SymmetricKey {
        let cacheKey = "\(password)\u{0}\(salt.base64EncodedString())"
        if let key = keyCacheQueue.sync(execute: { cachedKeys[cacheKey] }) {
            return key
        }

        let key = try deriveKey(password: password, salt: salt)
        keyCacheQueue.sync {
            cachedKeys[cacheKey] = key
        }
        return key
    }

    private static func deriveKey(password: String, salt: Data) throws -> SymmetricKey {
        guard let passwordData = password.data(using: .utf8) else {
            throw CryptoError.invalidPassword
        }

        var key = Data(count: keyBytes)
        let derivedKeyCount = key.count
        let result = key.withUnsafeMutableBytes { keyBytes in
            salt.withUnsafeBytes { saltBytes in
                passwordData.withUnsafeBytes { passwordBytes in
                    CCKeyDerivationPBKDF(
                        CCPBKDFAlgorithm(kCCPBKDF2),
                        passwordBytes.bindMemory(to: Int8.self).baseAddress,
                        passwordData.count,
                        saltBytes.bindMemory(to: UInt8.self).baseAddress,
                        salt.count,
                        CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
                        UInt32(pbkdf2Rounds),
                        keyBytes.bindMemory(to: UInt8.self).baseAddress,
                        derivedKeyCount
                    )
                }
            }
        }

        guard result == kCCSuccess else {
            throw CryptoError.keyDerivationFailed
        }
        return SymmetricKey(data: key)
    }

    private static func randomData(count: Int) -> Data {
        var data = Data(count: count)
        _ = data.withUnsafeMutableBytes {
            SecRandomCopyBytes(kSecRandomDefault, count, $0.bindMemory(to: UInt8.self).baseAddress!)
        }
        return data
    }
}

enum CryptoError: Error {
    case invalidPassword
    case invalidEnvelope
    case unsupportedEnvelope
    case keyDerivationFailed
}
